### PR TITLE
`createCaller()`-fn + simplify SSR helpers

### PR DESCRIPTION
**Pull request by KATT**
- create `router.createCaller()`
- make `router.invoke()`-private
- make ssr helpers a lot simpler

### CODE REVIEW

File: packages/server/test/index.test.tsx [hunk 1]
```diff
@@ -278,3 +278,48 @@ describe('integration tests', () => {
     });
   });
 });
+
+describe('call()', () => {
```
Comment (cyrus-za) [inline packages/server/test/index.test.tsx:282]:
```suggestion
describe('createCaller()', () => {
```

File: packages/server/test/index.test.tsx [hunk 2]
```diff
@@ -278,3 +278,48 @@ describe('integration tests', () => {
...
+
+  test('query()', async () => {
+    const data = await router.createCaller({}).query('q', 1);
+    expectTypeOf(data).toMatchTypeOf<{ input: number }>();
+  });
+  test('mutation()', async () => {
+    const data = await router.createCaller({}).mutation('m', 1);
+    expectTypeOf(data).toMatchTypeOf<{ input: number }>();
+  });
+  test('subscription()', async () => {
```
Comment (cyrus-za) [inline packages/server/test/index.test.tsx:320]: I dont think there's a use-case for subscription when on server-side (which is the whole point of this)

**Review by cyrus-za — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -429,6 +432,30 @@ export default MyApp;
...
+  return {
+    props: {
+      allPosts: trpc.dehydrate(),
+    },
+    revalidate: 1,
+  };
+}
+```
+</details>
+
```
Comment (KATT) [inline README.md:458]: @cyrus-za I've added you as a maintainer so you should be able to push to this branch - could you add more details here? I guess one would want the `inferServerSideProps` or whatever it's called from next

Comment (cyrus-za) [reply README.md:458]: You mean `InferGetStaticPropsType<typeof getStaticProps>` ?

File: packages/server/src/router.ts [hunk 4]
```diff
@@ -382,6 +382,31 @@ export class Router<
...
   }
 
+  public createCaller(
+    ctx: TContext,
+  ): {
+    query: inferHandlerFn<TQueries>;
+    mutation: inferHandlerFn<TMutations>;
+  } {
+    return {
+      query: (path, ...args) => {
```
Comment (cyrus-za) [inline packages/server/src/router.ts:392]:
```suggestion
      query: (path, input, ...args) => {
```

File: packages/server/src/router.ts [hunk 5]
```diff
@@ -382,6 +382,31 @@ export class Router<
...
+        return this.invoke({
+          target: 'queries',
+          ctx,
+          path,
+          input: args[0],
```
Comment (cyrus-za) [inline packages/server/src/router.ts:397]:
```suggestion
          input,
```

File: packages/server/src/router.ts [hunk 6]
```diff
@@ -382,6 +382,31 @@ export class Router<
...
+        }) as any;
+      },
+      mutation: (path, ...args) => {
```
Comment (cyrus-za) [inline packages/server/src/router.ts:400]:
```suggestion
      mutation: (path, input, ...args) => {
```

File: packages/server/src/router.ts [hunk 7]
```diff
@@ -382,6 +382,31 @@ export class Router<
...
+        return this.invoke({
+          target: 'mutations',
+          ctx,
+          path,
+          input: args[0],
```
Comment (cyrus-za) [inline packages/server/src/router.ts:405]:
```suggestion
          input,
```

**Review by cyrus-za — DISMISSED** [review]

**Comment by KATT** [comment]
@cyrus-za the       

```ts
query: (path, input) => {
```

doesn't work :) you'll see if you clone it, but it's because `input` is optional

File: README.md (hunk 3 above)
Comment (KATT) [reply README.md:458]: that

File: packages/react/src/createReactQueryHooks.ts [hunk 8]
```diff
@@ -206,59 +206,52 @@ export function createReactQueryHooks<
...
+      TPath extends keyof TQueries & string,
+      TProcedure extends TQueries[TPath]
+    >(
+      ...pathAndArgs: [path: TPath, ...args: inferHandlerInput<TProcedure>]
+    ) => {
+      const [path, input] = pathAndArgs;
+      const cacheKey = [path, input ?? null];
+
+      return queryClient.prefetchQuery(cacheKey, async () => {
+        const data = await caller.query(...pathAndArgs);
```
Comment (cyrus-za) [inline packages/react/src/createReactQueryHooks.ts:227]:
```suggestion
        return caller.query(...pathAndArgs);
```

File: packages/react/src/createReactQueryHooks.ts [hunk 9]
```diff
@@ -206,59 +206,52 @@ export function createReactQueryHooks<
...
 
-    await queryClient.prefetchQuery(cacheKey, async () => {
-      const data = await router.invoke({
-        target: 'queries',
-        ctx,
-        path,
-        input,
+        return data;
```
Comment (cyrus-za) [inline packages/react/src/createReactQueryHooks.ts:229]:
```suggestion
```

File: packages/react/src/createReactQueryHooks.ts [hunk 10]
```diff
@@ -206,59 +206,52 @@ export function createReactQueryHooks<
...
-      path: TPath;
-      ctx: TContext;
-      input: TInput;
-    },
-  ): Promise<void> => {
-    const input = opts.input ?? null;
-    const { path, ctx } = opts;
-    const cacheKey = [CACHE_PREFIX_INFINITE_QUERIES, path, input];
+      return queryClient.prefetchInfiniteQuery(cacheKey, async () => {
+        const data = await caller.query(...pathAndArgs);
```
Comment (cyrus-za) [inline packages/react/src/createReactQueryHooks.ts:243]:
```suggestion
        return caller.query(...pathAndArgs);
```

File: packages/react/src/createReactQueryHooks.ts [hunk 11]
```diff
@@ -206,59 +206,52 @@ export function createReactQueryHooks<
...
 
-    await queryClient.prefetchInfiniteQuery(cacheKey, async () => {
-      const data = await router.invoke({
-        target: 'queries',
-        ctx,
-        path,
-        input,
+        return data;
```
Comment (cyrus-za) [inline packages/react/src/createReactQueryHooks.ts:245]:
```suggestion
```

File: packages/react/src/createReactQueryHooks.ts [hunk 12]
```diff
@@ -206,59 +206,52 @@ export function createReactQueryHooks<
...
+    };
 
-      return data;
-    });
-  };
+    return {
+      caller,
+      prefetchQuery,
+      prefetchInfiniteQuery,
+    };
```
Comment (cyrus-za) [inline packages/react/src/createReactQueryHooks.ts:253]: I like this. One object for all the ssr methods, then the default client thing can expose hooks

File: packages/server/src/router.ts [hunk 13]
```diff
@@ -342,7 +350,7 @@ export class Router<
    * @throws RouteNotFoundError
    * @throws InputValidationError
    */
-  public async invoke(opts: {
+  private async invoke(opts: {
```
Comment (cyrus-za) [inline packages/server/src/router.ts:353]: 👍🏻

File: packages/server/src/router.ts (hunk 4 above)
Comment (cyrus-za) [reply packages/server/src/router.ts:382]: Why resolve this without leaving a comment?

File: packages/server/src/router.ts [hunk 14]
```diff
@@ -362,6 +370,41 @@ export class Router<
...
   }
 
+  public createCaller(
+    ctx: TContext,
+  ): {
+    query: inferHandlerFn<TQueries>;
+    mutation: inferHandlerFn<TMutations>;
+    subscription: inferHandlerFn<TSubscriptions>;
+  } {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
```
Comment (cyrus-za) [inline packages/server/src/router.ts:380]:
```suggestion
```
I dont think this is relevant anymore

File: packages/server/test/index.test.tsx [hunk 15]
```diff
@@ -29,34 +29,14 @@ test('mix query and mutation', async () => {
...
-      target: 'queries',
-      path: 'q2',
-      input: {
-        q2: 'hey',
-      },
-      ctx: {},
-    }),
-  ).toMatchInlineSnapshot(`"q2res"`);
+  expect(await caller.query('q2', { q2: 'hey' })).toMatchInlineSnapshot(
+    `"q2res"`,
```
Comment (cyrus-za) [inline packages/server/test/index.test.tsx:36]: should we not test the inputs getting sent through properly? Change the router above from
```ts
    .query('q2', {
      input: z.object({ q2: z.string() }),
      resolve() {
        return 'q2res';
      },
    })
```
to
```ts
    .query('q2', {
      input: z.object({ q2: z.string() }),
      resolve({ input }) {
        return `q2res: ${input.q2}`;
      },
    })
```

Then
```suggestion
    `"q2res: hey"`,
```

File: packages/server/test/index.test.tsx [hunk 16]
```diff
@@ -314,3 +288,55 @@ describe('integration tests', () => {
     });
   });
 });
+
+describe('createCall()', () => {
```
Comment (cyrus-za) [inline packages/server/test/index.test.tsx:292]:
```suggestion
describe('createCaller()', () => {
```

File: packages/server/test/index.test.tsx [hunk 17]
```diff
@@ -314,3 +288,55 @@ describe('integration tests', () => {
...
+            };
+          },
+        });
+      },
+    });
+
+  test('query()', async () => {
+    const data = await router.createCaller({}).query('q', 1);
+    expectTypeOf(data).toMatchTypeOf<{ input: number }>();
+    expect(data).toEqual({ input: 1 });
```
Comment (cyrus-za) [inline packages/server/test/index.test.tsx:325]: I guess the inputs get tested here. so the comments above aren't that important

File: packages/server/test/react.test.tsx [hunk 18]
```diff
@@ -440,14 +437,9 @@ test('useInfiniteQuery()', async () => {
...
+  const ssr = hooks.ssr(appRouter, {});
 
-  await hooks.prefetchInfiniteQueryOnServer(appRouter, {
-    path: 'paginatedPosts',
-    input: {
-      limit: 1,
-    },
-    ctx: {},
-  });
+  await ssr.prefetchInfiniteQuery('paginatedPosts', { limit: 1 });
```
Comment (cyrus-za) [inline packages/server/test/react.test.tsx:442]: test name does not match method

**Review by cyrus-za — CHANGES_REQUESTED** [review]

File: README.md [hunk 19]
```diff
@@ -397,43 +403,51 @@ import { trpc } from '../utils/trpc'
...
+  await ssr.prefetchInfiniteQuery('messages.list', {});
+  // or `await ssr.prefetchQuery('messages.list', {});`
+
   return {
     props: {
       dehydratedState: trpc.dehydrate(),
     },
-    revalidate: 1,
+    // revalidate: 1,
```
[the remark this answers is not in the document]
Comment (KATT) [reply README.md:416]: because it's just reference code

**Comment by KATT** [comment]
@all-contributors add @cyrus-za for review

**Pull request merged**